## Heroes of Pymoli

In [12]:
# Dependencies and Setup
import pandas as pd

# File to Load
data = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
data_df = pd.read_csv(data)

#data_df.head()

## Player Count


* Display the total number of players

In [13]:
#Find list of unique User Names for player count
total_players = len(data_df["SN"].unique())

#Create a dataframe for the total players
total_players = [{'Total Players':total_players}]
total_players_df = pd.DataFrame(total_players)
total_players_df

Total Players
0            576

## Purchasing Analysis

* Display the:
    * Number of Unique Items
    * Average Purchase Price
    * Total Number of Purchases
    * Total Revenue

In [14]:
#Find length of the list of unique items
unique_items = len(data_df["Item Name"].unique())

#Find average price of items using mean
avg_price= data_df["Price"].mean(axis=0)

#find total amount of purchases
total_purchase=len(data_df["Purchase ID"])

#sum the amount purchased
total_rev = data_df["Price"].sum(axis=0)

#create a dataframe with findings
purchasing_analysis = [{'Number of Unique Items': unique_items,
                        'Average Price': avg_price,
                        'Number of Purchases': total_purchase,
                        'Total Revenue': total_rev}]
purchasing_analysis_df = pd.DataFrame(purchasing_analysis)

#format to dollars
purchasing_analysis_df["Total Revenue"] = purchasing_analysis_df["Total Revenue"].map("${:,.2f}".format)
purchasing_analysis_df['Average Price'] = purchasing_analysis_df['Average Price'].map("${:,.2f}".format)
purchasing_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed

In [15]:
            #Collect a list of all male
male_data = data_df.loc[data_df["Gender"] == "Male"]
#Using GroupBy in order to separate data into fields according to "SN" users
male_sn = male_data.groupby(["SN","Gender"])
#find percentage of male
male_count = len(male_sn)
percent_male = (male_count/576)


            #Collect a list of all female
female_data = data_df.loc[data_df["Gender"] == "Female"]
#Using GroupBy in order to separate data into fields according to "SN" users
female_sn = female_data.groupby(["SN","Gender"])
#find percentage of female
female_count = len(female_sn)
percent_female = (female_count/576)


            #Collect a list of all other
other_data = data_df.loc[data_df["Gender"] == "Other / Non-Disclosed"]
#Using GroupBy in order to separate data into fields according to "SN" users
other_sn = other_data.groupby(["SN","Gender"])
#find percentage of "other"
other_count = len(other_sn)
percent_other = (other_count/576)

#create a dataframe with findings
gender_demographics = ({'': ['Male', 'Female', 'Other / Non-Disclosed'],
                        'Total Count': [male_count, female_count, other_count],
                        'Percentage of Players': [percent_male, percent_female, percent_other]})
gender_demographics_df = pd.DataFrame(gender_demographics).set_index("")

#format to percentage
gender_demographics_df['Percentage of Players'] = gender_demographics_df['Percentage of Players'].map("{:.2%}".format)
gender_demographics_df


Total Count Percentage of Players
                                                        
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

## Gender Purchasing Analysis

*Below each of the following are broken down by gender:
* Purchase Count
* Average Purchase Price
* Total Purchase Value
* Average Purchase Total per Person by Gender

In [16]:
# Purchase Count
male_count_df = male_sn.count()
male_purchase_count = male_count_df["Purchase ID"].sum()
            
female_count_df = female_sn.count()
female_purchase_count = female_count_df["Purchase ID"].sum()
            
other_count_df = other_sn.count()
other_purchase_count = other_count_df["Purchase ID"].sum()

#Groupby Gender to find averages
gender_df = data_df.groupby(["Gender"])

# Average Purchase Price
avg_price_gender = gender_df["Price"].mean().round(2)
avg_price_list = list(avg_price_gender)



# Average Purchase Total per Person by Male
each_purchase_male = male_sn["Price"].sum()
avg_total_purchase_male = each_purchase_male.mean()
#Total Purchase Value Male
total_purchase_male = each_purchase_male.sum()



# Average Purchase Total per Person by Female
each_purchase_female = female_sn["Price"].sum()
avg_total_purchase_female = each_purchase_female.mean()
# Total Purchase Value Female
total_purchase_female = each_purchase_female.sum()



# Average Purchase Total per Person by Other
each_purchase_other = other_sn["Price"].sum()
avg_total_purchase_other = each_purchase_other.mean()
# Total Purchase Value Other
total_purchase_other = each_purchase_other.sum()


#Create dataframe with findings
gender_purchasing_analysis = ({'': ["Male", "Female", "Other/ Non-Disclosed"],
                        'Purchase Count': [male_purchase_count, female_purchase_count, other_purchase_count],
                        'Average Purchase Price': [avg_price_list[0], avg_price_list[1], avg_price_list[2]],
                        'Total Purchase Value': [total_purchase_male, total_purchase_female, total_purchase_other],
                        'Avg Total Purchase per Person': [avg_total_purchase_male, avg_total_purchase_female, avg_total_purchase_other] })
gender_purchasing_analysis_df = pd.DataFrame(gender_purchasing_analysis).set_index("")

#format each column to appropriate values
gender_purchasing_analysis_df['Avg Total Purchase per Person'] = gender_purchasing_analysis_df['Avg Total Purchase per Person'].map("${:,.2f}".format)
gender_purchasing_analysis_df['Average Purchase Price'] = gender_purchasing_analysis_df['Average Purchase Price'].map("${:,.2f}".format)
gender_purchasing_analysis_df['Total Purchase Value'] = gender_purchasing_analysis_df['Total Purchase Value'].map("${:,.2f}".format)
gender_purchasing_analysis_df

Purchase Count Average Purchase Price  \
                                                              
Male                             652                  $3.20   
Female                           113                  $3.02   
Other/ Non-Disclosed              15                  $3.35   

                     Total Purchase Value Avg Total Purchase per Person  
                                                                         
Male                            $1,967.64                         $4.07  
Female                            $361.94                         $4.47  
Other/ Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Below each are broken into bins of 4 years
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Age Group

In [17]:
#Groupby individual and age to get uniue list of users
age_per_person = data_df.groupby(["SN","Age"])

#find total spent per person
total_spent_per_person = age_per_person["Price"].sum()
#find number of purchases they made
number_of_purchases_per_person= age_per_person["Price"].count()
#find average total spent per person
avg_purchase_per_person = total_spent_per_person/number_of_purchases_per_person

#create a dataframe to hold this data to merge later
summary_age_per_person_df = pd.DataFrame(avg_purchase_per_person)
summary_age_per_person_df["Players Count"] = (number_of_purchases_per_person)
summary_age_per_person_df["Total Purchase Value"] = total_spent_per_person

#get rid of groupby index
per_person_df = summary_age_per_person_df.reset_index()
#per_person_df

In [18]:
#Create bins in which to place players age
bins = [0,9,14,19,24,29,34,39,50] 
#create labels for these bins
group_ages = ["<10","10-14","15-19","20-24", "25-29", "30-34", "35-39","40+"]
#cut our "per_person" dataframe to see player count inside of the age groups
per_person_df["Age Ranges"] = pd.cut(per_person_df["Age"], bins, labels = group_ages)


#Groupby the labels of our bins
age_group = per_person_df.groupby(["Age Ranges"])

#Players Count
age_group_df = age_group[["Players Count"]].count()
#find percentage of players in each age group
age_group_df["Percentage of Players"] = (age_group_df/576)
#format to percentage
age_group_df["Percentage of Players"] = age_group_df["Percentage of Players"].map("{:.2%}".format)
age_group_df

Players Count Percentage of Players
Age Ranges                                     
<10                    17                 2.95%
10-14                  22                 3.82%
15-19                 107                18.58%
20-24                 258                44.79%
25-29                  77                13.37%
30-34                  52                 9.03%
35-39                  31                 5.38%
40+                    12                 2.08%

## Age Purchasing Analysis

In [19]:
#Cut our original data fram to count all purchases from repeat customers
data_df["Age Ranges"] = pd.cut(data_df["Age"], bins, labels = group_ages)

#rename column to use in our datafram as "Purchase Count"
data_df_renamed = data_df.rename(columns={"Purchase ID": "Purchase Count"})

#Group data by age groups created
age_ranges = data_df_renamed.groupby(["Age Ranges"])

# Purchase Count
age_ranges_df = age_ranges[["Purchase Count"]].count()
age_ranges[["Purchase Count"]].count()

# Average Purchase Price
age_ranges_df["Average Purchase Price"] = age_ranges["Price"].mean().map("${:,.2f}".format)

# Total Purchase Value
age_ranges_df["Total Purchase Value"] = age_ranges["Price"].sum()


#merge the two data frames to use player count for avg purchase total by person in each age range
age_purchasing_analysis = pd.merge(age_ranges_df,age_group_df, on="Age Ranges")

#Use purchase value and player count to find the average of total purchase per person
age_purchasing_analysis["Avg Purchase Total per Person"] = age_purchasing_analysis["Total Purchase Value"]/age_purchasing_analysis["Players Count"]
age_purchasing_analysis["Avg Purchase Total per Person"]=age_purchasing_analysis["Avg Purchase Total per Person"].map("${:,.2f}".format)

#DROP the values/columns not asked for - you can see these in previous chart
age_purchasing_analysis = age_purchasing_analysis.drop("Percentage of Players", axis=1)
age_purchasing_analysis = age_purchasing_analysis.drop("Players Count", axis=1)
age_purchasing_analysis

Purchase Count Average Purchase Price  Total Purchase Value  \
Age Ranges                                                                
<10                     23                  $3.35                 77.13   
10-14                   28                  $2.96                 82.78   
15-19                  136                  $3.04                412.89   
20-24                  365                  $3.05               1114.06   
25-29                  101                  $2.90                293.00   
30-34                   73                  $2.93                214.00   
35-39                   41                  $3.60                147.67   
40+                     13                  $2.94                 38.24   

           Avg Purchase Total per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value

In [20]:
#categorize by Total Purchase value in descending order - show top 5
top_spender_summary_df = per_person_df.sort_values(by=["Total Purchase Value"], ascending=False).head(5)
top_spender_summary_df = pd.DataFrame(top_spender_summary_df)

#remove age index and delete column
top_spender_summary_df=top_spender_summary_df.set_index("SN")
top_spender_summary_df=top_spender_summary_df.drop("Age", axis=1)
top_spender_renamed_df=top_spender_summary_df.rename(columns={"Price": "Avg Purchase Price", "Players Count": "Purchase Count"})

#format price and purchase value as dollars
top_spender_renamed_df["Avg Purchase Price"]=top_spender_renamed_df["Avg Purchase Price"].map("${:,.2f}".format)
top_spender_renamed_df["Total Purchase Value"]=top_spender_renamed_df["Total Purchase Value"].map("${:,.2f}".format)
top_spender_rearranged_df = top_spender_renamed_df[["Purchase Count", "Avg Purchase Price","Total Purchase Value"]]
top_spender_rearranged_df

Purchase Count Avg Purchase Price Total Purchase Value
SN                                                                 
Lisosia93                 5              $3.79               $18.96
Idastidru52               4              $3.86               $15.45
Chamjask73                3              $4.61               $13.83
Iral74                    4              $3.40               $13.62
Iskadarya95               3              $4.37               $13.10

## Most Popular Items

* Identify the 5 most popular items by purchase count, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [21]:
#Group Items by ID and Item Name
item_group = data_df.groupby(["Item ID","Item Name"])

#Count the number of purchases
total_number_sold = item_group[["Purchase ID"]].count()
total_number_sold["Total Purchase Value"] = item_group["Price"].sum()


total_number_sold["Price"] = total_number_sold["Total Purchase Value"]/total_number_sold["Purchase ID"]

descending_purchase_count = total_number_sold.sort_values(by=["Purchase ID"], ascending=False)


#Rearrange columns
descending_purchase_count = descending_purchase_count[["Purchase ID", "Price","Total Purchase Value"]]

descending_purchase_count["Price"]=descending_purchase_count["Price"].map("${:,.2f}".format)
descending_purchase_count["Total Purchase Value"]=descending_purchase_count["Total Purchase Value"].map("${:,.2f}".format)

#RENAME Purchase ID to Purchase Count
descending_purchase_count=descending_purchase_count.rename(columns={"Purchase ID": "Purchase Count"})


top_five_purchased_items = descending_purchase_count.head(5)
top_five_purchased_items



Purchase Count  Price  \
Item ID Item Name                                                             
92      Final Critic                                              13  $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm              12  $4.23   
145     Fiery Glass Crusader                                       9  $4.58   
132     Persuasion                                                 9  $3.22   
108     Extraction, Quickblade Of Trembling Hands                  9  $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items


* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [22]:
descending_purchase_value = total_number_sold.sort_values(by=["Total Purchase Value"], ascending=False)

#Rearrange columns
descending_purchase_value = descending_purchase_value[["Purchase ID", "Price","Total Purchase Value"]]

descending_purchase_value["Price"]=descending_purchase_value["Price"].map("${:,.2f}".format)
descending_purchase_value["Total Purchase Value"]=descending_purchase_value["Total Purchase Value"].map("${:,.2f}".format)

#RENAME Purchase ID to Purchase Count
descending_purchase_value=descending_purchase_value.rename(columns={"Purchase ID": "Purchase Count"})

top_five_profitable_items = descending_purchase_value.head(5)
top_five_profitable_items

Purchase Count  Price  \
Item ID Item Name                                                             
92      Final Critic                                              13  $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm              12  $4.23   
82      Nirvana                                                    9  $4.90   
145     Fiery Glass Crusader                                       9  $4.58   
103     Singed Scalpel                                             8  $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Observations and Trends

*Three observable trends based of data:

#1. Over 84% of their players are male. This is a male dominant game, which also means that majority of their purchase value comes from male purchases

#2. On average each purchase is around 3 dollara. This makes sense because with around 780 transactions we see that there were over 2,300 dollars (3x780) in purchases. Now we can assume how many more transactions we would need to acheive a higher Purchase Value Goal. Even top spenders are spending 3-4 dollars on each transaction.

#3. Majority of the users are in their 20's. Specifially early 20's takes up almost 45% of the user base.